source/common/buffer/buffer_impl.cc

```diff
@@ -12,99 +13,319 @@
+
+  void* data() override {
+    return static_cast<uint8_t*>(const_cast<void*>(fragment_.data())) + data_;
+  }
+
+  uint64_t dataSize() const override { return fragment_.size() - data_; }
+
+  void drain(uint64_t size) override {
+    ASSERT(data_ + size <= fragment_.size());
+    data_ += size;
+  }
+
+  uint64_t reservableSize() const override { return 0; }
+
+  Reservation reserve(uint64_t) override { return {nullptr, 0}; }
+
+  bool commit(const Reservation&, uint64_t) override { return false; }
+
+private:
+  BufferFragment& fragment_;
+
+  /**
+   * Offset in bytes from the start of the fragment to the start of the Data section
+   * (increases with each call to drain()).
+   */
+  uint64_t data_;
+};
+
+void OwnedImpl::append(const void* data, uint64_t size) {
+  const char* src = static_cast<const char*>(data);
+  while (size != 0) {
+    if (slices_.empty() || slices_.back()->reservableSize() == 0) {
+      slices_.emplace_back(std::make_unique<OwnedBufferSlice>(size));
+    }
+    BufferSlice::Reservation reservation = slices_.back()->reserve(size);
+    ASSERT(reservation.first != nullptr);
+    ASSERT(reservation.second != 0);
+    memcpy(reservation.first, src, reservation.second);
+    slices_.back()->commit(reservation, reservation.second);
+    src += reservation.second;
+    size -= reservation.second;
+  }
 }
 
-void OwnedImpl::add(absl::string_view data) {
-  evbuffer_add(buffer_.get(), data.data(), data.size());
+void OwnedImpl::add(const void* data, uint64_t size) { append(data, size); }
+
+void OwnedImpl::addBufferFragment(BufferFragment& fragment) {
+  slices_.emplace_back(std::make_unique<UnownedBufferSlice>(fragment));
 }
 
+void OwnedImpl::add(absl::string_view data) { append(data.data(), data.size()); }
+
 void OwnedImpl::add(const Instance& data) {
   uint64_t num_slices = data.getRawSlices(nullptr, 0);
   STACK_ARRAY(slices, RawSlice, num_slices);
   data.getRawSlices(slices.begin(), num_slices);
   for (const RawSlice& slice : slices) {
-    add(slice.mem_, slice.len_);
+    append(slice.mem_, slice.len_);
   }
 }
 
 void OwnedImpl::prepend(absl::string_view data) {
-  evbuffer_prepend(buffer_.get(), data.data(), data.size());
+  slices_.emplace_front(std::make_unique<OwnedBufferSlice>(data.data(), data.size()));
 }
 
 void OwnedImpl::prepend(Instance& data) {
-  int rc =
-      evbuffer_prepend_buffer(buffer_.get(), static_cast<LibEventInstance&>(data).buffer().get());
-  ASSERT(rc == 0);
-  ASSERT(data.length() == 0);
-  static_cast<LibEventInstance&>(data).postProcess();
+  OwnedImpl& other = static_cast<OwnedImpl&>(data);
+  while (!other.slices_.empty()) {
+    slices_.emplace_front(std::move(other.slices_.back()));
+    other.slices_.pop_back();
+  }
+  other.postProcess();
 }
 
 void OwnedImpl::commit(RawSlice* iovecs, uint64_t num_iovecs) {
-  int rc =
-      evbuffer_commit_space(buffer_.get(), reinterpret_cast<evbuffer_iovec*>(iovecs), num_iovecs);
-  ASSERT(rc == 0);
+  int64_t i = num_iovecs;
+  i--;
```

Comment:
My rationale for doing the subtraction in signed space is that subtracting in unsigned space can cause an underflow (it's valid for `num_iovecs` to be zero). The only reason the overflow doesn't break anything is that the result is converted to a signed value of the same size (uint64_t to int64_t), so `i` ends up being `-1` as expected. But if somebody ever changes `num_iovecs` to, say, `unsigned` in the future, `i` will end up having a large positive value instead of `-1`. But I can at least replace:
```
int64_t i = num_iovecs;
i--;
```
with a cleaner equivalent:
```
int64_t i = static_cast<int64_t>(num_iovecs) - 1;
```